### PR TITLE
New reconnection handling implementation. Now we will try to reconnec…

### PR DESCRIPTION
**Pull request by stopal-r7**
…t with a fully configurable backoff mechanism that supports both fibonacci and exponential.

Buffer will buffer log events from the time the disconnection was detected until it is reconnected. After a successful reconnection, buffer will be flushed.

Broken test after defaulting to secure connection is fixed.

this should fix: #101 #110

### CODE REVIEW

**Comment by stopal-r7** [comment]
cc: @dtracey-r7
@danilosterrapid7 @pquinn-r7 please review

File: src/node_modules/logger.js [hunk 1]
```diff
@@ -308,94 +351,81 @@ export default class Logger extends Writable {
...
+		});
 
-				this.emit('connected');
-			});
+		this[$reconnection].on('connection timeout', () => {
+			// we owe a lot to inactivity timeout handling with regards to clearing unwanted opened connections hanging around.
+			DEBUG.log("Socket was inactive for a period of time. Destroying.");
+			this.closeConnection();
+			delete this[$connection];
+			this.emit("timed out");
```
Comment (pquinn-r7) [inline src/node_modules/logger.js]: is anything listening for this emit?

File: src/node_modules/logger.js [hunk 2]
```diff
@@ -131,27 +136,58 @@ export default class Logger extends Writable {
...
+
+		this[$reconnect] = reconnect_core(function () {
+			var connection;
+			if (secure) {
+				connection = tls.connect.apply(tls, arguments, () => {
+					if (!connection.authorized) {
+						DEBUG.log("Secure connection -TLS- is not authorized.");
+						const errMsg = connection.authorizationError;
+						this.emit(new LogentriesError(text.authError(errMsg)));
+					} else if (tls && tls.CleartextStream && connection instanceof tls.CleartextStream)
```
Comment (pquinn-r7) [inline src/node_modules/logger.js]: Stick with curly brackets even if not necessary.

File: src/node_modules/logger.js [hunk 3]
```diff
@@ -476,15 +527,31 @@ export default class Logger extends Writable {
...
 	}
 
 	@nonConfigurable
+	get inactivityTimeout() {
+		return this[$timeout];
+	}
+
+	set inactivityTimeout(val) {
+		if (Number.isInteger(val) && val >= 0)
+			this[$timeout] = parseInt(val);
```
Comment (pquinn-r7) [inline src/node_modules/logger.js]: same here

File: src/node_modules/logger.js [hunk 4]
```diff
@@ -476,15 +527,31 @@ export default class Logger extends Writable {
...
+
+		if (!_.isNumber(this[$timeout]))
+			this[$timeout] = defaults.inactivityTimeout;
```
Comment (pquinn-r7) [inline src/node_modules/logger.js]: and here

File: src/node_modules/logger.js [hunk 5]
```diff
@@ -278,19 +315,25 @@ export default class Logger extends Writable {
...
+
 	@unwritable
 	@enumerable
 	closeConnection() {
-		if (!this[$connection]) return;
+		DEBUG.log("Closing retry mechanism along with its connection.");
 
-		this[$connection].then(conn => conn.end());
+		if (!this[$reconnection])
+			return;
```
Comment (pquinn-r7) [inline src/node_modules/logger.js]: and here

File: package.json [hunk 6]
```diff
@@ -47,7 +47,8 @@
     "codependency": "0.1.4",
     "json-stringify-safe": "5.0.1",
     "lodash": "3.9.3",
-    "semver": "^5.1.0"
+    "semver": "^5.1.0",
```
Comment (pquinn-r7) [inline package.json]: Can we keep this all to exact versions?

File: src/node_modules/logger.js [hunk 7]
```diff
@@ -1,54 +1,51 @@
...
+const $timestamp                = Symbol('timestamp property');
+const $token                    = Symbol('token property');
+const $toLevel                  = Symbol('toLevel method');
+const $withLevel                = Symbol('withLevel property');
+const $withStack                = Symbol('withStack property');
 
 // PATTERNS ////////////////////////////////////////////////////////////////////
 
-const newline = /\n/g;
+const newline      = /\n/g;
```
Comment (pquinn-r7) [inline src/node_modules/logger.js]: random spacing

File: src/node_modules/logger.js (hunk 1 above)
Comment (stopal-r7) [inline src/node_modules/logger.js]: Not in le_node, but this is for exposing timeout event to the application that uses le_node.

Comment (danilosterrapid7) [inline src/node_modules/logger.js]: Would be nice if you had the event name as an exported constant. It makes life easier for those who wants to consume it.

File: src/node_modules/logger.js [hunk 8]
```diff
@@ -71,7 +68,7 @@ const requirePeer = codependency.register(module);
 
 const nonConfigurable = (target, name, descriptor) => {
 	descriptor.configurable = false;
-	descriptor.enumerable = true;
+	descriptor.enumerable   = true;
```
Comment (danilosterrapid7) [inline src/node_modules/logger.js]: disuniform spacing

File: src/node_modules/logger.js (hunk 7 above)
Comment (stopal-r7) [inline src/node_modules/logger.js]: This spacing is forced and for to align it with the following assignments. Makes the code more readable.

File: src/node_modules/logger.js [hunk 9]
```diff
@@ -131,27 +136,58 @@ export default class Logger extends Writable {
...
+			} else {
+				connection = net.connect.apply(null, arguments);
+			}
 
-		this.flatten = opts.flatten;
-		this.flattenArrays =
-			'flattenArrays' in opts ? opts.flattenArrays : opts.flatten;
+			connection.setTimeout(opts.inactivityTimeout || defaults.inactivityTimeout);
+			// Expose the connection’s drain event
+			connection.on('drain', () => {
```
Comment (danilosterrapid7) [inline src/node_modules/logger.js]: It would be nice having all the events as exported constants.

File: src/node_modules/logger.js (hunk 8 above)
Comment (stopal-r7) [inline src/node_modules/logger.js]: This is to align assignment operator with previous assignment line. More readable code.

File: src/node_modules/logger.js [hunk 10]
```diff
@@ -131,27 +136,58 @@ export default class Logger extends Writable {
...
+					setImmediate(() => this.emit('connection drain'));
```
Comment (danilosterrapid7) [inline src/node_modules/logger.js]: It would be nice having all the events as exported constants.

File: src/node_modules/logger.js [hunk 11]
```diff
@@ -308,94 +351,81 @@ export default class Logger extends Writable {
...
-					} else if (tls && tls.CleartextStream &&
-						connection instanceof tls.CleartextStream)
-					// Your node is ancient!
-						this.emit('connect');
+			this[$reconnection].on('connect', function (con) {
+				DEBUG.log("Connected");
+				this.emit('connected');
+				con.on('timeout', () => {
+					//we listen to timeout event and close connection if it happens. see below.
+					this.emit("connection timeout");
```
Comment (danilosterrapid7) [inline src/node_modules/logger.js]: It would be nice having all the events as exported constants.

File: src/node_modules/logger.js [hunk 12]
```diff
@@ -308,94 +351,81 @@ export default class Logger extends Writable {
...
 			});
 
-			// Close will emit whether it was intended or error-caused:
-			connection.on('close', () => {
-				delete this[$connection];
-
-				this.emit('disconnected');
+			this[$reconnection].on('disconnect', function () {
+				DEBUG.log("Disconnected.");
+				this.emit("disconnected");
```
Comment (danilosterrapid7) [inline src/node_modules/logger.js]: here as well

File: src/node_modules/logger.js (hunk 8 above)
Comment (danilosterrapid7) [inline src/node_modules/logger.js]: I really do not think is more readable, especially, when we have a considerable gap between variable and operator. 

But, this is my point of view...

File: src/node_modules/logger.js (hunk 11 above)
Comment (stopal-r7) [inline src/node_modules/logger.js]: This event is emitted to $reconnection and used internally but I know what you mean. On it.

**Comment by danilosterrapid7** [comment]
+1
Agreed to apply guide styles in a separate PR.

**Comment by pquinn-r7** [comment]
+1

**Pull request merged**